transports/quic/src/provider/tokio.rs

```diff
@@ -18,67 +18,26 @@
 // FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 // DEALINGS IN THE SOFTWARE.
 
-use futures::{ready, Future};
+use futures::Future;
 use std::{
     io,
-    net::SocketAddr,
     task::{Context, Poll},
 };
-use tokio::{io::ReadBuf, net::UdpSocket};
 
 use crate::GenTransport;
 
 /// Transport with [`tokio`] runtime.
 pub type Transport = GenTransport<Provider>;
 
-/// Provider for reading / writing to a sockets and spawning
-/// tasks using [`tokio`].
-pub struct Provider {
-    socket: UdpSocket,
-    socket_recv_buffer: Vec<u8>,
-    next_packet_out: Option<(Vec<u8>, SocketAddr)>,
-}
+/// Provider for quinn runtime and spawning tasks using [`tokio`].
+pub struct Provider;
 
 impl super::Provider for Provider {
     type IfWatcher = if_watch::tokio::IfWatcher;
+    type Runtime = quinn::TokioRuntime;
```

Comment:
In the previous implementation, we took great care that none of the `quinn-proto` types leak into the public API to make sure updates to the library don't break our public API.

I believe referencing the type in the assoicated type here makes it part of the public API. Can we work around this?

Reply:
Frankly speaking I don't know how to fix it. Any ideas?

Reply:
Is there a use-case for anyone to implement `Provider` on their own custom type? If not, how about making the `Provider` trait non-public? That way we don't expose `quinn` types in our public interface.

What do you think @thomaseizinger and @kpp?

Reply:
As an alternative, we could hide the underlying `quinn` types entirely by using an `enum` instead of a `trait`. Again users would not be able to to _provide_ their own `Provider` implementation.

Happy to propose a patch in case we decide on one of the two (sealed `trait` or `enum`).

Reply:
I think you can still refer to types in a sealed trait  meaning a breaking version bump still has a breaking semver impact on our crate.

We can't add support for a runtime unless `quinn` does as far as I am aware. As a result, the user can't really bring their own runtime anyway.

Whatever solution we use, I'd like `quinn` to be completely hidden from the public API.

Reply:
> Whatever solution we use, I'd like `quinn` to be completely hidden from the public API.

I'll have a look at how `quinn` is structured internally and propose something.

Reply:
> As an alternative, we could hide the underlying quinn types entirely by using an enum

We cannot do this because we need to unpack the underlying runtime. We need to patch quinn: either make quinn::Endpoint::new_with_runtime public or implement quinn::Runtime for Box<dyn quinn::Runtime>.

Reply:
`impl Runtime for Box<Runtime>` worked. Here is the patch:

<details>
  <summary>libp2p patch</summary>

  
  ```diff
diff --git a/transports/quic/src/provider.rs b/transports/quic/src/provider.rs
index 8b616257..cd2e95dd 100644
--- a/transports/quic/src/provider.rs
+++ b/transports/quic/src/provider.rs
@@ -30,13 +30,32 @@ pub mod async_std;
 #[cfg(feature = "tokio")]
 pub mod tokio;
 
+pub enum Runtime {
+    #[cfg(feature = "tokio")]
+    Tokio(quinn::TokioRuntime),
+    #[cfg(feature = "async-std")]
+    AsyncStd(quinn::AsyncStdRuntime),
+    Dummy,
+}
+
+impl Runtime {
+    pub(crate) fn get(self) -> Option<Box<dyn quinn::Runtime>> {
+        match self {
+            #[cfg(feature = "tokio")]
+            Runtime::Tokio(runtime) => Some(Box::new(runtime)),
+            #[cfg(feature = "async-std")]
+            Runtime::AsyncStd(runtime) => Some(Box::new(runtime)),
+            _ => None
+        }
+    }
+}
+
 /// Provider for a corresponding quinn runtime and spawning tasks.
 pub trait Provider: Unpin + Send + Sized + 'static {
     type IfWatcher: Unpin + Send;
-    type Runtime: quinn::Runtime;
 
     /// Run the corresponding runtime
-    fn runtime() -> Self::Runtime;
+    fn runtime() -> Runtime;
 
     /// Run the given future in the background until it ends.
     ///
diff --git a/transports/quic/src/provider/async_std.rs b/transports/quic/src/provider/async_std.rs
index d5dc2cdf..b08f1033 100644
--- a/transports/quic/src/provider/async_std.rs
+++ b/transports/quic/src/provider/async_std.rs
@@ -35,10 +35,9 @@ pub struct Provider;
 
 impl super::Provider for Provider {
     type IfWatcher = if_watch::smol::IfWatcher;
-    type Runtime = quinn::AsyncStdRuntime;
 
-    fn runtime() -> Self::Runtime {
-        quinn::AsyncStdRuntime
+    fn runtime() -> super::Runtime {
+        super::Runtime::AsyncStd(quinn::AsyncStdRuntime)
     }
 
     fn spawn(future: impl Future<Output = ()> + Send + 'static) {
diff --git a/transports/quic/src/provider/tokio.rs b/transports/quic/src/provider/tokio.rs
index adeb0906..1016c89d 100644
--- a/transports/quic/src/provider/tokio.rs
+++ b/transports/quic/src/provider/tokio.rs
@@ -34,10 +34,9 @@ pub struct Provider;
 
 impl super::Provider for Provider {
     type IfWatcher = if_watch::tokio::IfWatcher;
-    type Runtime = quinn::TokioRuntime;
 
-    fn runtime() -> Self::Runtime {
-        quinn::TokioRuntime
+    fn runtime() -> super::Runtime {
+        super::Runtime::Tokio(quinn::TokioRuntime)
     }
 
     fn spawn(future: impl Future<Output = ()> + Send + 'static) {
diff --git a/transports/quic/src/transport.rs b/transports/quic/src/transport.rs
index 76f8da5f..c2518020 100644
--- a/transports/quic/src/transport.rs
+++ b/transports/quic/src/transport.rs
@@ -96,7 +96,8 @@ impl<P: Provider> GenTransport<P> {
         socket_addr: SocketAddr,
     ) -> Result<quinn::Endpoint, Error> {
         let socket = UdpSocket::bind(socket_addr)?;
-        let endpoint = quinn::Endpoint::new(endpoint_config, server_config, socket, P::runtime())?;
+        let runtime = P::runtime().get().unwrap();
+        let endpoint = quinn::Endpoint::new(endpoint_config, server_config, socket, runtime)?;
         Ok(endpoint)
     }
 }
  
  ```
  
</details>

I will open a PR in quinn repo.

Reply:
Done